// webflow-api/dist/Webflow.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _isomorphicFetch = _interopRequireDefault(require("isomorphic-fetch"));

var _qs = _interopRequireDefault(require("qs"));

var _utils = require("./utils");

var _ResponseWrapper = _interopRequireDefault(require("./ResponseWrapper"));

var _WebflowError = _interopRequireWildcard(require("./WebflowError"));

var _excluded = ["collectionId"],
    _excluded2 = ["collectionId", "itemId"],
    _excluded3 = ["collectionId", "itemId"],
    _excluded4 = ["collectionId", "itemIds"],
    _excluded5 = ["collectionId", "itemIds"],
    _excluded6 = ["siteId", "userId"],
    _excluded7 = ["siteId"];

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }

var DEFAULT_ENDPOINT = "https://api.webflow.com";

var buildMeta = function buildMeta(res) {
  if (!res || !res.headers) {
    return {};
  }

  return {
    rateLimit: {
      limit: parseInt(res.headers.get("x-ratelimit-limit"), 10),
      remaining: parseInt(res.headers.get("x-ratelimit-remaining"), 10)
    }
  };
};

var responseHandler = function responseHandler(res) {
  return res.json()["catch"](function (err) {
    return (// Catch unexpected server errors where json isn't sent and rewrite
      // with proper class (WebflowError)
      Promise.reject(new _WebflowError["default"](err))
    );
  }).then(function (body) {
    if (res.status >= 400) {
      var errOpts = {
        code: body.code,
        msg: body.msg,
        _meta: buildMeta(res)
      };

      if (body.problems && body.problems.length > 0) {
        errOpts.problems = body.problems;
      }

      var errMsg = body && body.err ? body.err : "Unknown error occured";
      var err = new _WebflowError["default"](errMsg);
      return Promise.reject(Object.assign(err, errOpts));
    }

    body._meta = buildMeta(res); // eslint-disable-line no-param-reassign

    return body;
  });
};

var Webflow = /*#__PURE__*/function () {
  function Webflow() {
    var _this = this;

    var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
        _ref$endpoint = _ref.endpoint,
        endpoint = _ref$endpoint === void 0 ? DEFAULT_ENDPOINT : _ref$endpoint,
        token = _ref.token,
        _ref$version = _ref.version,
        version = _ref$version === void 0 ? "1.0.0" : _ref$version;

    _classCallCheck(this, Webflow);

    if (!token) throw (0, _WebflowError.buildRequiredArgError)("token");
    this.responseWrapper = new _ResponseWrapper["default"](this);
    this.endpoint = endpoint;
    this.token = token;
    this.headers = {
      Accept: "application/json",
      Authorization: "Bearer ".concat(token),
      "accept-version": version,
      "Content-Type": "application/json"
    };

    this.authenticatedFetch = function (method, path, data, query) {
      var queryString = query && !(0, _utils.isObjectEmpty)(query) ? "?".concat(_qs["default"].stringify(query)) : "";
      var uri = "".concat(_this.endpoint).concat(path).concat(queryString);
      var opts = {
        method: method,
        headers: _this.headers,
        mode: "cors"
      };

      if (data) {
        opts.body = JSON.stringify(data);
      }

      return (0, _isomorphicFetch["default"])(uri, opts).then(responseHandler);
    };
  } // Generic HTTP request handlers


  _createClass(Webflow, [{
    key: "get",
    value: function get(path) {
      var query = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      return this.authenticatedFetch("GET", path, null, query);
    }
  }, {
    key: "post",
    value: function post(path, data) {
      var query = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
      return this.authenticatedFetch("POST", path, data, query);
    }
  }, {
    key: "put",
    value: function put(path, data) {
      var query = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
      return this.authenticatedFetch("PUT", path, data, query);
    }
  }, {
    key: "patch",
    value: function patch(path, data) {
      var query = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
      return this.authenticatedFetch("PATCH", path, data, query);
    }
  }, {
    key: "delete",
    value: function _delete(path, data) {
      var query = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
      return this.authenticatedFetch("DELETE", path, data, query);
    } // Meta

  }, {
    key: "info",
    value: function info() {
      var query = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      return this.get("/info", query);
    } // Sites

  }, {
    key: "sites",
    value: function sites() {
      var _this2 = this;

      var query = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
      return this.get("/sites", query).then(function (sites) {
        return sites.map(function (site) {
          return _this2.responseWrapper.site(site);
        });
      });
    }
  }, {
    key: "site",
    value: function site(_ref2) {
      var _this3 = this;

      var siteId = _ref2.siteId;
      var query = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      if (!siteId) return Promise.reject((0, _WebflowError.buildRequiredArgError)("siteId"));
      return this.get("/sites/".concat(siteId), query).then(function (site) {
        return _this3.responseWrapper.site(site);
      });
    }
  }, {
    key: "publishSite",
    value: function publishSite(_ref3) {
      var siteId = _ref3.siteId,
          domains = _ref3.domains;
      if (!siteId) return Promise.reject((0, _WebflowError.buildRequiredArgError)("siteId"));
      if (!domains) return Promise.reject((0, _WebflowError.buildRequiredArgError)("domains"));
      return this.post("/sites/".concat(siteId, "/publish"), {
        domains: domains
      });
    } // Domains

  }, {
    key: "domains",
    value: function domains(_ref4) {
      var _this4 = this;

      var siteId = _ref4.siteId;
      if (!siteId) return Promise.reject((0, _WebflowError.buildRequiredArgError)("siteId"));
      return this.get("/sites/".concat(siteId, "/domains")).then(function (domains) {
        return domains.map(function (domain) {
          return _this4.responseWrapper.domain(domain, siteId);
        });
      });
    } // Collections

  }, {
    key: "collections",
    value: function collections(_ref5) {
      var _this5 = this;

      var siteId = _ref5.siteId;
      var query = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      if (!siteId) return Promise.reject((0, _WebflowError.buildRequiredArgError)("siteId"));
      return this.get("/sites/".concat(siteId, "/collections"), query).then(function (collections) {
        return collections.map(function (collection) {
          return _this5.responseWrapper.collection(collection);
        });
      });
    }
  }, {
    key: "collection",
    value: function collection(_ref6) {
      var _this6 = this;

      var collectionId = _ref6.collectionId;
      var query = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      if (!collectionId) return Promise.reject((0, _WebflowError.buildRequiredArgError)("collectionId"));
      return this.get("/collections/".concat(collectionId), query).then(function (collection) {
        return _this6.responseWrapper.collection(collection);
      });
    } // Items

  }, {
    key: "items",
    value: function items(_ref7) {
      var _this7 = this;

      var collectionId = _ref7.collectionId;
      var query = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      if (!collectionId) return Promise.reject((0, _WebflowError.buildRequiredArgError)("collectionId"));
      return this.get("/collections/".concat(collectionId, "/items"), query).then(function (res) {
        return _objectSpread(_objectSpread({}, res), {}, {
          items: res.items.map(function (item) {
            return _this7.responseWrapper.item(item, collectionId);
          })
        });
      });
    }
  }, {
    key: "item",
    value: function item(_ref8) {
      var _this8 = this;

      var collectionId = _ref8.collectionId,
          itemId = _ref8.itemId;
      var query = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      if (!collectionId) return Promise.reject((0, _WebflowError.buildRequiredArgError)("collectionId"));
      if (!itemId) return Promise.reject((0, _WebflowError.buildRequiredArgError)("itemId"));
      return this.get("/collections/".concat(collectionId, "/items/").concat(itemId), query).then(function (res) {
        return _this8.responseWrapper.item(res.items[0], collectionId);
      });
    }
  }, {
    key: "createItem",
    value: function createItem(_ref9) {
      var _this9 = this;

      var collectionId = _ref9.collectionId,
          data = _objectWithoutProperties(_ref9, _excluded);

      var query = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      if (!collectionId) return Promise.reject((0, _WebflowError.buildRequiredArgError)("collectionId"));
      return this.post("/collections/".concat(collectionId, "/items"), data, query).then(function (item) {
        return _this9.responseWrapper.item(item, collectionId);
      });
    }
  }, {
    key: "updateItem",
    value: function updateItem(_ref10) {
      var collectionId = _ref10.collectionId,
          itemId = _ref10.itemId,
          data = _objectWithoutProperties(_ref10, _excluded2);

      var query = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      if (!collectionId) return Promise.reject((0, _WebflowError.buildRequiredArgError)("collectionId"));
      if (!itemId) return Promise.reject((0, _WebflowError.buildRequiredArgError)("itemId"));
      return this.put("/collections/".concat(collectionId, "/items/").concat(itemId), data, query);
    }
  }, {
    key: "removeItem",
    value: function removeItem(_ref11) {
      var collectionId = _ref11.collectionId,
          itemId = _ref11.itemId;
      var query = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      if (!collectionId) return Promise.reject((0, _WebflowError.buildRequiredArgError)("collectionId"));
      if (!itemId) return Promise.reject((0, _WebflowError.buildRequiredArgError)("itemId"));
      return this["delete"]("/collections/".concat(collectionId, "/items/").concat(itemId), null, query);
    }
  }, {
    key: "patchItem",
    value: function patchItem(_ref12) {
      var collectionId = _ref12.collectionId,
          itemId = _ref12.itemId,
          data = _objectWithoutProperties(_ref12, _excluded3);

      var query = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      if (!collectionId) return Promise.reject((0, _WebflowError.buildRequiredArgError)("collectionId"));
      if (!itemId) return Promise.reject((0, _WebflowError.buildRequiredArgError)("itemId"));
      return this.patch("/collections/".concat(collectionId, "/items/").concat(itemId), data, query);
    }
  }, {
    key: "deleteItems",
    value: function deleteItems(_ref13) {
      var collectionId = _ref13.collectionId,
          itemIds = _ref13.itemIds,
          data = _objectWithoutProperties(_ref13, _excluded4);

      var query = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      if (!collectionId) return Promise.reject((0, _WebflowError.buildRequiredArgError)("collectionId"));
      if (!itemIds) return Promise.reject((0, _WebflowError.buildRequiredArgError)("itemIds"));
      return this["delete"]("/collections/".concat(collectionId, "/items"), _objectSpread(_objectSpread({}, data), {}, {
        itemIds: itemIds
      }), query);
    }
  }, {
    key: "publishItems",
    value: function publishItems(_ref14) {
      var collectionId = _ref14.collectionId,
          itemIds = _ref14.itemIds,
          data = _objectWithoutProperties(_ref14, _excluded5);

      var query = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      if (!collectionId) return Promise.reject((0, _WebflowError.buildRequiredArgError)("collectionId"));
      if (!itemIds) return Promise.reject((0, _WebflowError.buildRequiredArgError)("itemIds"));
      return this.put("/collections/".concat(collectionId, "/items/publish"), _objectSpread(_objectSpread({}, data), {}, {
        itemIds: itemIds
      }), query);
    } // Users

  }, {
    key: "users",
    value: function users(_ref15) {
      var _this10 = this;

      var siteId = _ref15.siteId;
      var query = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      if (!siteId) return Promise.reject((0, _WebflowError.buildRequiredArgError)("siteId"));
      return this.get("/sites/".concat(siteId, "/users"), query).then(function (res) {
        return res.users.map(function (user) {
          return _this10.responseWrapper.user(user);
        });
      });
    }
  }, {
    key: "user",
    value: function user(_ref16) {
      var _this11 = this;

      var siteId = _ref16.siteId,
          userId = _ref16.userId;
      var query = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      if (!siteId) return Promise.reject((0, _WebflowError.buildRequiredArgError)("siteId"));
      if (!userId) return Promise.reject((0, _WebflowError.buildRequiredArgError)("userId"));
      return this.get("/sites/".concat(siteId, "/users/").concat(userId), query).then(function (user) {
        return _this11.responseWrapper.user(user, siteId);
      });
    }
  }, {
    key: "updateUser",
    value: function updateUser(_ref17) {
      var siteId = _ref17.siteId,
          userId = _ref17.userId,
          data = _objectWithoutProperties(_ref17, _excluded6);

      var query = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      if (!siteId) return Promise.reject((0, _WebflowError.buildRequiredArgError)("siteId"));
      if (!userId) return Promise.reject((0, _WebflowError.buildRequiredArgError)("userId"));
      return this.patch("/sites/".concat(siteId, "/users/").concat(userId), data, query);
    }
  }, {
    key: "inviteUser",
    value: function inviteUser(_ref18) {
      var _this12 = this;

      var siteId = _ref18.siteId,
          email = _ref18.email;
      var query = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      if (!siteId) return Promise.reject((0, _WebflowError.buildRequiredArgError)("siteId"));
      if (!email) return Promise.reject((0, _WebflowError.buildRequiredArgError)("email"));
      return this.post("/sites/".concat(siteId, "/users/invite"), {
        email: email
      }, query).then(function (user) {
        return _this12.responseWrapper.user(user, siteId);
      });
    }
  }, {
    key: "removeUser",
    value: function removeUser(_ref19) {
      var siteId = _ref19.siteId,
          userId = _ref19.userId;
      var query = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      if (!siteId) return Promise.reject((0, _WebflowError.buildRequiredArgError)("siteId"));
      if (!userId) return Promise.reject((0, _WebflowError.buildRequiredArgError)("userId"));
      return this["delete"]("/sites/".concat(siteId, "/users/").concat(userId), null, query);
    } // Webhooks

  }, {
    key: "webhooks",
    value: function webhooks(_ref20) {
      var _this13 = this;

      var siteId = _ref20.siteId;
      var query = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      if (!siteId) return Promise.reject((0, _WebflowError.buildRequiredArgError)("siteId"));
      return this.get("/sites/".concat(siteId, "/webhooks"), query).then(function (webhooks) {
        return webhooks.map(function (webhook) {
          return _this13.responseWrapper.webhook(webhook, siteId);
        });
      });
    }
  }, {
    key: "webhook",
    value: function webhook(_ref21) {
      var _this14 = this;

      var siteId = _ref21.siteId,
          webhookId = _ref21.webhookId;
      var query = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      if (!siteId) return Promise.reject((0, _WebflowError.buildRequiredArgError)("siteId"));
      if (!webhookId) return Promise.reject((0, _WebflowError.buildRequiredArgError)("webhookId"));
      return this.get("/sites/".concat(siteId, "/webhooks/").concat(webhookId), query).then(function (webhook) {
        return _this14.responseWrapper.webhook(webhook, siteId);
      });
    }
  }, {
    key: "createWebhook",
    value: function createWebhook(_ref22) {
      var _this15 = this;

      var siteId = _ref22.siteId,
          data = _objectWithoutProperties(_ref22, _excluded7);

      var query = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      if (!siteId) return Promise.reject((0, _WebflowError.buildRequiredArgError)("siteId"));
      return this.post("/sites/".concat(siteId, "/webhooks"), data, query).then(function (webhook) {
        return _this15.responseWrapper.webhook(webhook, siteId);
      });
    }
  }, {
    key: "removeWebhook",
    value: function removeWebhook(_ref23) {
      var siteId = _ref23.siteId,
          webhookId = _ref23.webhookId;
      var query = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      if (!siteId) return Promise.reject((0, _WebflowError.buildRequiredArgError)("siteId"));
      if (!webhookId) return Promise.reject((0, _WebflowError.buildRequiredArgError)("webhookId"));
      return this["delete"]("/sites/".concat(siteId, "/webhooks/").concat(webhookId), null, query);
    }
  }]);

  return Webflow;
}();

exports["default"] = Webflow;
